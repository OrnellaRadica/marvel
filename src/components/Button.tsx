import Link from "next/link";
import { ComponentProps } from "react";
import styled, { css } from "styled-components";

interface BaseProps {
  variant: "primary" | "secondary";
  children: React.ReactNode;
}

type LinkProps = Omit<ComponentProps<typeof Link>, "href"> & {
  href: string;
};

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  href?: undefined;
}

type Props = BaseProps & (ButtonProps | LinkProps);

function primaryColorStyles() {
  return css`
    background-color: #444;
    color: #fff;
    &:hover {
      background-color: #1b1a1a;
    }
  `;
}

function secondaryColorStyles() {
  return css`
    background-color: #fff;
    color: #444;
    outline: 2px solid #444;

    &:hover {
      background-color: #e4e4e4;
    }
  `;
}

function buttonBaseStyles(props: Props) {
  const { variant } = props;

  return css`
    padding: 16px 24px;
    text-decoration: none;
    border-radius: 4px;
    display: flex;
    gap: 16px;
    font-weight: 600;
    font-size: 18px;
    letter-spacing: -0.16px;
    border: none;
    width: fit-content;
    ${() => {
      switch (variant) {
        case "primary":
          return primaryColorStyles();
        case "secondary":
          return secondaryColorStyles();
        default:
          primaryColorStyles();
      }
    }}
  `;
}

const StyledButton = styled.button<BaseProps & ButtonProps>(buttonBaseStyles);
const StyledAnchor = styled.a<BaseProps & LinkProps>(buttonBaseStyles);
const StyledLink = styled(Link)<BaseProps & LinkProps>(buttonBaseStyles);

export function Button(props: Props) {
  if (typeof props.href !== "string") {
    return <StyledButton type="button" {...props} />;
  }

  if (typeof props.href === "string" && props.href.startsWith("/")) {
    return <StyledLink {...props} />;
  }
  // In this case it's an external link
  return <StyledAnchor {...props} />;
}
