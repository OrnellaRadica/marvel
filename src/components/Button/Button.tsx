import Link from "next/link";
import { ComponentProps } from "react";
import styled, { css } from "styled-components";

interface BaseProps {
  variant?: "primary" | "secondary";
  children: React.ReactNode;
}

type LinkProps = Omit<ComponentProps<typeof Link>, "href"> & {
  href: string;
};

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  href?: undefined;
}

type Props = BaseProps & (ButtonProps | LinkProps);

function primaryColorStyles() {
  return css`
    background-color: #444;
    color: #fff;
    &:hover {
      background-color: #1b1a1a;
    }
  `;
}

function secondaryColorStyles() {
  return css`
    background-color: #fff;
    color: #444;
    outline: 2px solid #444;

    &:hover {
      background-color: #e4e4e4;
    }
  `;
}

interface ButtonBaseStylesProps {
  $variant?: string;
}

function buttonBaseStyles(props: ButtonBaseStylesProps) {
  const { $variant } = props;

  return css`
    padding: 16px 24px;
    text-decoration: none;
    border-radius: 4px;
    display: flex;
    gap: 16px;
    font-weight: 600;
    font-size: 18px;
    letter-spacing: -0.16px;
    border: none;
    width: fit-content;
    ${() => {
      switch ($variant) {
        case "primary":
          return primaryColorStyles();
        case "secondary":
          return secondaryColorStyles();
        default:
          primaryColorStyles();
      }
    }}
  `;
}

const StyledButton = styled.button<ButtonBaseStylesProps & ButtonProps>(
  buttonBaseStyles
);
const StyledAnchor = styled.a<ButtonBaseStylesProps & LinkProps>(
  buttonBaseStyles
);
const StyledLink = styled(Link)<ButtonBaseStylesProps & LinkProps>(
  buttonBaseStyles
);

export function Button(props: Props) {
  const { variant, ...rest } = props;
  if (typeof rest.href !== "string") {
    return <StyledButton type="button" {...rest} $variant={variant} />;
  }

  if (typeof rest.href === "string" && rest.href.startsWith("/")) {
    return <StyledLink {...rest} $variant={variant} />;
  }
  // In this case it's an external link
  return <StyledAnchor {...rest} $variant={variant} />;
}
