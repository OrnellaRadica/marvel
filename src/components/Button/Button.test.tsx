import renderer from "react-test-renderer";
import { Button } from "./Button";

describe("Button", () => {
  it("Should return a button", () => {
    const button = renderer.create(<Button>Example button</Button>).toJSON();
    expect(button).toMatchSnapshot();
  });

  it("Should return a button with secondary prop", () => {
    const button = renderer
      .create(<Button variant="secondary">Example button</Button>)
      .toJSON();
    expect(button).toMatchSnapshot();
  });

  it("Should return a button with type 'submit'", () => {
    const button = renderer
      .create(<Button type="submit">Example button</Button>)
      .toJSON();
    expect(button).toMatchSnapshot();
  });

  it("Should return a link NextJs component", () => {
    const button = renderer
      .create(<Button href="/">Example button</Button>)
      .toJSON();
    expect(button).toMatchSnapshot();
  });

  it("Should return a link to an external href", () => {
    const button = renderer
      .create(<Button href="https://www.marvel.com/">Example button</Button>)
      .toJSON();
    expect(button).toMatchSnapshot();
  });

  it("Should return a link to an external href in a new tab", () => {
    const button = renderer
      .create(<Button href="https://www.marvel.com/" target="_blank">Example button</Button>)
      .toJSON();
    expect(button).toMatchSnapshot();
  });
});
