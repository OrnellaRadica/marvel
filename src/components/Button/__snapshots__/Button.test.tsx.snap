// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should return a button 1`] = `
<button
  className="sc-16ed12da-0 eVbrCW"
  type="button"
>
  Example button
</button>
`;

exports[`Button Should return a button with secondary prop 1`] = `
<button
  className="sc-16ed12da-0 bwtpxf"
  type="button"
>
  Example button
</button>
`;

exports[`Button Should return a button with type 'submit' 1`] = `
<button
  className="sc-16ed12da-0 eVbrCW"
  type="submit"
>
  Example button
</button>
`;

exports[`Button Should return a link NextJs component 1`] = `
<a
  className="sc-16ed12da-2 ksRcFg"
  href="/"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  Example button
</a>
`;

exports[`Button Should return a link to an external href 1`] = `
<a
  className="sc-16ed12da-1 dPvqNf"
  href="https://www.marvel.com/"
>
  Example button
</a>
`;

exports[`Button Should return a link to an external href in a new tab 1`] = `
<a
  className="sc-16ed12da-1 dPvqNf"
  href="https://www.marvel.com/"
  target="_blank"
>
  Example button
</a>
`;
